### PR TITLE
Stream pausing

### PR DESCRIPTION
**Pull request by Jmgr**
This PR proposes an API to “pause” a stream, that is, deactivate its partitions. Publication to one of these partitions using the Liftbridge API “resumes” the stream, restarting its partitions.

## Context

Our use-case will include having a significant number of streams, but only a small fraction of those will be active at any given point in time (sparse streams).

In case of a server shutdown we would like to restart only active streams, not paused ones. The activity would record stream pausing and resuming, so that on server startup only the active streams would be restored.

The aims of this PR are twofold:

- inactive streams are paused so as to avoid the memory and CPU cost of maintaining stream activity and replication when there are no messages;

- by passing pause and resume events to a meta-event stream, we can schedule activity for a consumer pool more effectively.

This PR is therefore linked to the activity stream PR (https://github.com/liftbridge-io/liftbridge/pull/169).

## Implementation

The implementation proposed in this PR provides a new `PauseStream` API that allows clients to request that a stream be paused. Pausing a steam is very similar to closing a stream; that is, closing all of its partitions. Closing a partition, in turn, means that the server is no longer a leader or a follower for it and that its commit log is also closed. A “paused” flag is also set on the stream’s partitions. This allows us to spare CPU, memory usage and close unused file handles. One could also imagine an auto-pausing feature for each stream, where after a certain duration without any message publication a stream and its partition would be automatically paused. This would have to be added in a future PR however.

Resuming a stream occurs when a message is published to one of its partitions. All partition are then re-created using the same parameters, including the same data directory. The message is then published as usual.

### CODE REVIEW

File: server/api.go [hunk 1]
```diff
@@ -167,6 +185,28 @@ func (a *apiServer) Publish(ctx context.Context, req *client.PublishRequest) (
 	}
 	a.logger.Debugf("api: Publish [subject=%s]", subject)
 
+	stream := a.metadata.GetStream(req.Stream)
+	if stream == nil {
```
Comment (tylertreat) [inline server/api.go]: It just occurred to me that this won't work for [`PublishToSubject`](https://github.com/liftbridge-io/go-liftbridge/blob/master/client.go#L181-L192) which publishes Liftbridge messages directly to a NATS subject. That's probably OK since the expectation is that `Publish` operates at a stream level while `PublishToSubject` doesn't.

In the case of `PublishToSubject`, `Stream` will be `nil` on the `PublishRequest` since `Subject` is set instead. This now has me wondering if `Publish` and `PublishToSubject` should just be separate explicit endpoints.

File: server/api.go [hunk 2]
```diff
@@ -167,6 +185,28 @@ func (a *apiServer) Publish(ctx context.Context, req *client.PublishRequest) (
...
+	}
+	partition, ok := stream.partitions[0] // Only check the first partition for now
+	if !ok {
+		return nil, status.Error(codes.NotFound, fmt.Sprintf("No partition in stream: %s", req.Stream))
+	}
+	if partition.paused {
+		// This partition (and thus, this stream) is paused.
+		// We need to re-create the stream before being able to use it.
+		protoPart := partition.Partition
+		for i := 0; i < len(stream.partitions); i++ {
```
Comment (tylertreat) [inline server/api.go]: See discussion on https://github.com/liftbridge-io/liftbridge-api/pull/18 about pausing/resuming at the partition level.

File: server/metadata.go [hunk 3]
```diff
@@ -561,6 +603,7 @@ func (m *metadataAPI) addPartition(protoPartition *proto.Partition, recovered bo
 	if err != nil {
 		return nil, err
 	}
+	partition.paused = false
 	st.partitions[protoPartition.Id] = partition
```
Comment (tylertreat) [inline server/metadata.go]: I'd have to check if there are other references, but we'll want to ensure that the old, paused partition that is replaced gets properly garbage collected.

Comment (tylertreat) [reply server/metadata.go]: Alternatively, rather than using `CreatePartition` to unpause partitions, I wonder if it would make more sense to add a new Raft operation `UnpausePartition` that's able to resume an existing paused partition rather than creating a new one in its place? I could be convinced either way, just want to make sure we don't have any unforeseen memory leaks.

File: server/server_test.go [hunk 4]
```diff
@@ -1320,3 +1320,65 @@ func TestPropagatedShrinkExpandISR(t *testing.T) {
...
+	num := 1
+	for i := 0; i < num; i++ {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_, err = client.Publish(ctx, name, []byte("hello"))
+		require.NoError(t, err)
+	}
+
+	// Force log clean.
+	forceLogClean(t, subject, name, s1)
```
Comment (tylertreat) [inline server/server_test.go]: Why is this needed?

File: server/server_test.go [hunk 5]
```diff
@@ -1320,3 +1320,65 @@ func TestPropagatedShrinkExpandISR(t *testing.T) {
...
+	subject := "foo"
+	err = client.CreateStream(context.Background(), subject, name)
+	require.NoError(t, err)
+
+	err = client.PauseStream(context.Background(), name)
+	require.NoError(t, err)
+
+	// Publish some messages.
+	num := 1
+	for i := 0; i < num; i++ {
```
Comment (tylertreat) [inline server/server_test.go]: Can we get rid of the `for` loop since we're just publishing one message?

File: server/server_test.go [hunk 6]
```diff
@@ -1320,3 +1320,65 @@ func TestPropagatedShrinkExpandISR(t *testing.T) {
...
+		defer cancel()
+		_, err = client.Publish(ctx, name, []byte("hello"))
+		require.NoError(t, err)
+	}
+
+	// Force log clean.
+	forceLogClean(t, subject, name, s1)
+
+	// The first message read back should be the creation of the activity stream
+	// partition.
```
Comment (tylertreat) [inline server/server_test.go]: Was the copy/pasted from the activity stream PR by mistake?

Comment (Jmgr) [reply server/server_test.go]: Ah yes, I was a bit too hasty here. Should be fixed in 7025d22.

File: server/server_test.go (hunk 5 above)
Comment (Jmgr) [reply server/server_test.go]: Should be fixed in 7025d22.

File: server/server_test.go (hunk 4 above)
Comment (Jmgr) [reply server/server_test.go]: Should be fixed in 7025d22.

File: server/api.go (hunk 1 above)
Comment (Jmgr) [reply server/api.go]: Indeed. I've added a check for an empty stream name in 4a99036.

I do personally find it a bit counter intuitive to have two functions that are very similarly named but have a very different set of constraints/limitations.

File: server/metadata.go (hunk 3 above)
Comment (Jmgr) [reply server/metadata.go]: I have used `CreatePartition` here because it seemed simpler to re-create the partition using the same data path, as you suggested. What would be the steps involved in resuming the partition without re-creating it? Would re-starting the commit log and calling `startLeadingOrFollowing` be enough?

Comment (tylertreat) [reply server/metadata.go]: I think using `CreatePartition` is fine here.

File: server/api.go [hunk 7]
```diff
@@ -208,6 +231,34 @@ func (a *apiServer) Publish(ctx context.Context, req *client.PublishRequest) (
...
+	if !ok {
+		status.Error(codes.NotFound, fmt.Sprintf("No partition in stream: %s", req.Stream))
+	}
+	if partition.paused {
+		// This partition (and thus, this stream) is paused.
+		// We need to re-create the stream before being able to use it.
+		protoPart := partition.Partition
+		for i := 0; i < len(stream.partitions); i++ {
+			if e := a.metadata.CreatePartition(ctx, &proto.CreatePartitionOp{
+				Partition: protoPart,
```
Comment (tylertreat) [inline server/api.go]: It looks like this is creating the same partition over and over since it's reusing the `protoPart` from `stream.partitions[0]`. This should use the correct partition:

```go
for _, partition := range stream.partitions {
        if e := a.metadata.CreatePartition(ctx, &proto.CreatePartitionOp{
                Partition: partition.Partition,
        }); e != nil {
                ...
        }
}
```

File: server/api.go [hunk 8]
```diff
@@ -208,6 +231,34 @@ func (a *apiServer) Publish(ctx context.Context, req *client.PublishRequest) (
...
+			}); e != nil {
+				a.logger.Errorf("api: Failed to resume stream partition %d: %v", i, e.Err())
+				e.Err()
```
Comment (tylertreat) [inline server/api.go]: I think this is supposed to be a `return`?

File: server/api.go (hunk 7 above)
Comment (Jmgr) [reply server/api.go]: Ha, yes, good catch. Should be fixed in 1bb7320.

File: server/api.go (hunk 8 above)
Comment (Jmgr) [reply server/api.go]: Should be fixed in 1bb7320.

File: server/api.go (hunk 2 above)
Comment (Jmgr) [reply server/api.go]: Implemented in a01f499.

**Comment by tylertreat** [comment]
Should pausing survive restarts? I.e. if a stream is paused and the server is restarted, should the stream continue to be paused? My thought is yes.

**Comment by Jmgr** [comment]
> Should pausing survive restarts? I.e. if a stream is paused and the server is restarted, should the stream continue to be paused? My thought is yes.

Our idea for this was to use the activity stream to determine on restart if a partition has been paused.

**Comment by tylertreat** [comment]
I'm investigating a possible memory leak but will work on getting this merged in.
